lib/ads/client_side_ad_manager.js

```diff
@@ -97,6 +97,38 @@ shaka.ads.ClientSideAdManager = class {
     this.adsLoader_.requestAds(imaRequest);
   }
 
+  /**
+   * Start ads after checking autoplay
+   * @param {google.ima.AdsManager} imaAdsManager
+   * @private
+   */
+  async startAds_(imaAdsManager) {
+    const AutoPlay = new shaka.media.AutoPlay();
+
+    // IMA won't crash if autoplay is allowed, so we
+    // start the Ad Manager with no user interaction.
+    // If playback restrictions are in place, we wait for the video to be played
+    const {result} = await AutoPlay.canAutoPlay();
+    if (result === true) {
+      shaka.log.info('Autoplay allowed. Starting ads.');
+
+      if (imaAdsManager && imaAdsManager.start) {
+        imaAdsManager.start();
+      }
+    } else {
+      this.video_.addEventListener('play', () => {
+        if (imaAdsManager && imaAdsManager.start) {
+          imaAdsManager.start();
+        }
+      }, {once: true});
```

Comment:
Is there any reason you couldn't just do this unconditionally?  If autoplay is working, the 'play' event will be called when playback starts.  It seems like we don't really need autoplay detection at all.

Reply:
I think I've had some prerolls starting while the video is still buffering, so I guess it'd make for a smoother experience if we didn't wait for the "play" event. But I'll test that with different ads and videos to see how it goes.

Reply:
Let us know what you find.  Thanks!

Reply:
Sorry for the delay. I saved some URLs of a case in which I was able to reproduce this buffering issue.
These are the [manifest URL](https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8) and the [ad tag URL](https://pubads.g.doubleclick.net/gampad/ads?iu=/21775744923/external/single_ad_samples&sz=640x480&cust_params=sample_ct%3Dlinear&ciu_szs=300x250%2C728x90&gdfp_req=1&output=vast&unviewed_position_start=1&env=vp&impl=s&correlator=)

So what happens is: the spinner you see in the screenshot is actually from the content buffer, not from the ad. The ad has loaded faster than the content, so it plays before its buffering ends:

![image](https://user-images.githubusercontent.com/82478213/198411131-5f7b78dc-d84a-47e6-9756-96e725aa7637.png)

In cases where you have a 5-second ad this might be even more common. Plus, waiting for the content to play first might take away some of the user experience.

#

Waiting for the content to play, I've noticed a "flash" of content before the ad starts, like this:

https://user-images.githubusercontent.com/82478213/198412699-e6bbe5cb-62d6-432c-8816-5b31ea00db6f.mp4

The duration of the "flash" obviously depends on the user's connection, CDN speed, etc., but it is a reproducible case.

Reply:
I think the resolution will really depend on how much you think this negatively affects the experience. If it's fine like that, then we just listen for the play event; otherwise, the autoplay check is a better option.

Reply:
I think it's fine, personally.  An application can always choose not to use autoplay if it's a problem.  I'm also weighing that against the complexity of the code to detect autoplay.  What do you think?

Reply:
In theory, we just play the smallest video possible in a try/catch block, and that's our verification. Still more complex than just adding an event listener, though.
I guess I'll remove the autoplay detection for now and, if the new behaviour ever becomes an issue, we can just resort to the autoplay stuff again.